device/transport/amqp/test/_amqp_test.js

```diff
@@ -4,246 +4,1345 @@
+            transport.setOptions(badOptions);
+          }, ReferenceError, '');
+        });
+      });
+
+      /*Tests_SRS_NODE_DEVICE_AMQP_06_002: [If `done` has been specified the `setOptions` method shall call the `done` callback with no arguments when successful.]*/
+      it('calls the done callback with no arguments', function (done) {
+        transport.setOptions(testOptions, done);
+      });
+
+      /*Tests_SRS_NODE_DEVICE_AMQP_06_003: [`setOptions` should not throw if `done` has not been specified.]*/
+      it('does not throw if `done` is not specified', function () {
+        assert.doesNotThrow(function () {
+          transport.setOptions({});
+        });
+      });
     });
+  });
 
-    /*Tests_SRS_NODE_DEVICE_AMQP_06_003: [`setOptions` should not throw if `done` has not been specified.]*/
-    it('does not throw if `done` is not specified', function() {
-      var transport = new Amqp(configWithSAS);
-      assert.doesNotThrow(function() {
-        transport.setOptions({});
+  describe('D2C', function () {
+    describe('#sendEvent', function () {
+      /*Tests_SRS_NODE_DEVICE_AMQP_16_024: [The `sendEvent` method shall connect and authenticate the transport if necessary.]*/
+      it('automatically connects the transport if necessary', function (testCallback) {
+        transport.sendEvent(new Message('test'), function () {
+          assert(fakeBaseClient.connect.calledOnce);
+          testCallback();
+        });
+      });
+
+      it('forwards the error if connecting fails while trying to send a message', function (testCallback) {
+        var fakeError = new Error('failed to connect');
+        fakeBaseClient.connect = sinon.stub().callsArgWith(2, fakeError);
+
+        transport.sendEvent(new Message('test'), function (err) {
+          assert(fakeBaseClient.connect.calledOnce);
+          assert.strictEqual(err.amqpError, fakeError);
+          testCallback();
+        });
+      });
+
+      it('sends the message even if called while the transport is connecting', function (testCallback) {
+        var connectCallback;
+        fakeBaseClient.connect = sinon.stub().callsFake(function (uri, options, done) {
+          connectCallback = done;
+        });
+
+        transport.connect(function () {});
+
+        transport.sendEvent(new Message('test'), function () {
+          assert(fakeBaseClient.connect.calledOnce);
+          testCallback();
+        });
+
+        connectCallback(null, new results.Connected());
+      });
+
+      it('sends the message even if called while the transport is authenticating', function (testCallback) {
+        var authCallback;
+        fakeBaseClient.putToken = sinon.stub().callsFake(function (audience, token, done) {
+          authCallback = done;
+        });
+
+        transport.connect(function () {});
+
+        transport.sendEvent(new Message('test'), function () {
+          assert(fakeBaseClient.connect.calledOnce);
+          testCallback();
+        });
```

Comment:
Sometimes these things are just hard to follow.

Reply:
feel free to suggest a better pattern. adding a comment in the meantime.

---
In reply to: [141136711](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141136711) [](ancestors = 141136711)